keps/sig-storage/1472/README.md

```diff
@@ -0,0 +1,1349 @@
+
+* **How can a rollout fail? Can it impact already running workloads?**
+
+* **What specific metrics should inform a rollback?**
+
+* **Were upgrade and rollback tested? Was upgrade->downgrade->upgrade path tested?**
+
+* **Is the rollout accompanied by any deprecations and/or removals of features,
+  APIs, fields of API types, flags, etc.?**
+
+### Monitoring requirements
+
+Will be added before the transition to beta.
+
+* **How can an operator determine if the feature is in use by workloads?**
+
+* **What are the SLIs (Service Level Indicators) an operator can use to
+  determine the health of the service?**
+
+* **What are the reasonable SLOs (Service Level Objectives) for the above SLIs?**
+
+* **Are there any missing metrics that would be useful to have to improve
+  observability if this feature?**
+
+### Dependencies
+
+Will be added before the transition to beta.
+
+* **Does this feature depend on any specific services running in the cluster?**
+
+### Scalability
+
+Will be added before the transition to beta.
+
+* **Will enabling / using this feature result in any new API calls?**
+
+* **Will enabling / using this feature result in introducing new API types?**
+
+* **Will enabling / using this feature result in any new calls to cloud
+  provider?**
+
+* **Will enabling / using this feature result in increasing size or count
+  of the existing API objects?**
+
+* **Will enabling / using this feature result in increasing time taken by any
+  operations covered by [existing SLIs/SLOs][]?**
+
+* **Will enabling / using this feature result in non-negligible increase of
+  resource usage (CPU, RAM, disk, IO, ...) in any components?**
+
+### Troubleshooting
+
+Will be added before the transition to beta.
+
+* **How does this feature react if the API server and/or etcd is unavailable?**
+
+* **What are other known failure modes?**
+
+* **What steps should be taken if SLOs are not being met to determine the problem?**
+
+[supported limits]: https://git.k8s.io/community//sig-scalability/configs-and-limits/thresholds.md
+[existing SLIs/SLOs]: https://git.k8s.io/community/sig-scalability/slos/slos.md#kubernetes-slisslos
+
+
+## Implementation History
+
+- Kubernetes 1.19: alpha (tentative)
+
+
+## Drawbacks
+
+The attempt to define and implement a generic solution may end up with
+something that isn't capable enough in practice because it does not
+know enough about specific limitations of individual CSI drivers.
+
+At the moment, storage vendor can already achieve the same goals
+entirely without changes in Kubernetes or Kubernetes-CSI, it's just a
+lot of work (see the [TopoLVM
+design](https://github.com/cybozu-go/topolvm/blob/master/docs/design.md#diagram))
+- custom node and controller sidecars
+- scheduler extender (probably slower than a builtin scheduler
+  callback)
+
+
+## Alternatives
+
+### CSI drivers without topology support
+
+To simplify the implementation of external-provisioner, [topology
```

Comment:
I'm confused. Topology is a capability in CSI so drivers have to opt-in to support this.

Reply:
A certain @msau42 said that we can assume that CSI drivers support topology when they want to use storage capacity tracking :stuck_out_tongue_closed_eyes: 

I added this text in commit 86cf87dc7a2e58fcfec9034180d14ef6bab29fcb and removed some external-provisioner modes in response to your comment in https://github.com/kubernetes/enhancements/pull/1353/files#r369745671:
> I'm not sure we should have so many different modes of operation. This will make maintaining the provisioner very complex.
>
> I think we should try to simplify things and require that the plugin supports Topology in order to use this feature, and only have two modes of operation "local" and
"not local".

Reply:
Yes that part makes sense. I'm confused by the later sentence that talks about a driver implementing topology when they don't need it. What is the use case for a driver wanting the capacity feature when they don't support topology? The scheduler is not going to make any different decisions if capacity is equal across all nodes.

Reply:
Ah, now I see what you mean. Indeed, what the heck was I thinking of here? I don't remember. Let me remove the second sentence and if we find a use-case for it again, we can add back something better.